### ISSUE TITLE
framework: consider caching on-screen animated image frames

### ISSUE THREAD

**Issue by yjbanov** [flutter/flutter#83027]
It seems the `Image` widget does not cache frames from animated widgets, causing the image codec to run on every frame. In my experiments this doubles the amount of frame time assuming image animation as the only activity (e.g. in the repro below it goes from ~1ms to ~2ms on desktop). Perhaps the framework should cache previously decoded frames for images that are known to be visible on screen.

## Steps to Reproduce

1. Use the repro from https://github.com/flutter/flutter/issues/78223
2. `flutter run -d chrome --profile`
3. Record a profile using Chrome's "Performance" tab

Here's a sample profile. Notice that the GIF codec is exercised. This is on a screen with no user interaction, and the only animation is a couple of GIF images animating.

![gif-animation-profile](https://user-images.githubusercontent.com/211513/119015049-68fd8e80-b94d-11eb-80bd-699f3835d58c.PNG)

## Alternatives

Another way to optimize this is to have the framework call `Codec.dispose` when the animated image is no longer needed. The engine could keep the frames around until the codec is disposed of. A downside is that this approach would regress existing usages of `Codec` that do not call `dispose` explicitly.

**Comment by goderbauer** [comment]
This is currently working as implemented: The framework's ImageCache caches ImageStreamCompleters and the MultiFrameImageStreamCompleter is properly cached there. However, the MultiFrameImageStreamCompleter itself does not cache individual frames. Those are always decoded on demand for each frame.

We used to have some frame caching on the engine side, but that has been removed a while ago because it was the cause of many OOM bugs, discussion for that is in https://github.com/flutter/flutter/issues/26081. From the discussion there it sounds like this behavior is in line with e.g. what chrome is doing as well: Prioritizing memory (less caching) over CPU (re-decoding repeating frames), https://github.com/flutter/flutter/issues/24835#issuecomment-443296475.

So in summary, I am not convinced that adding such a cache to the framework is a good idea given the context.

**Comment by goderbauer** [comment]
Having said that, if one really wanted to (and is willing to take on and manage the OOM risk properly), one could implement a custom `MultiFrameImageStreamCompleter` based on the framework one that caches the frames instead of disposing them after they have been displayed.

**Comment by yjbanov** [comment]
Thanks for the explainers. Closing as this is WAI and has a workaround.
